File: demo/windows/Release/src/LinkDisk/network/TcpServer.java
package LinkDisk.network;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.FileOutputStream;
import java.net.ServerSocket;
import java.net.Socket;
import java.io.File;
import java.io.FileInputStream;
import java.nio.file.Files;
import java.nio.charset.StandardCharsets;
import java.nio.file.StandardCopyOption;
import java.text.SimpleDateFormat;
import java.util.Arrays;
import java.util.Date;

public class TcpServer {

    private static AuthManager authManager = new AuthManager();

    private static AuthCallback authCallback;

    private static ReceiveCallback receiveCallback;

    private static ServerSocket serverSocket;

    private static boolean isRunning = false;

    public interface AuthCallback {
        boolean onAuthRequest(String ip);
    }

    public interface ReceiveCallback {

        void onFileReceiveStart(
                String clientIp,
                String fileName,
                String savePath,
                long fileSize);

        void onFileReceiveProgress(
                String clientIp,
                String fileName,
                String savePath,
                long fileSize,
                long receivedBytes,
                int progress);

        void onFileReceived(
                String clientIp,
                String fileName,
                String savePath,
                long fileSize);
    }

    private static String formatFileSize(long size) {
        double value = size;

        if (size < 1000) {
            return size + " B";
        }

        value = value / 1000;
        if (value < 1000) {
            return String.format(java.util.Locale.US, "%.2f KB", value);
        }

        value = value / 1000;
        if (value < 1000) {
            return String.format(java.util.Locale.US, "%.2f MB", value);
        }

        value = value / 1000;
        return String.format(java.util.Locale.US, "%.2f GB", value);
    }

    public static void startServer(AuthCallback callback) {
        startServer(callback, null);
    }

    public static void startServer(AuthCallback callback, ReceiveCallback receiveCallbackParam) {
        if (isRunning) {
            return;
        }

        authCallback = callback;
        receiveCallback = receiveCallbackParam;
        isRunning = true;

        new File(AppSettings.getReceiveDir()).mkdirs();

        new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    serverSocket = new ServerSocket(6000);
                    System.out.println("TCP服务器已启动，端口：6000");

                    while (isRunning) {

                        Socket socket = serverSocket.accept();

                        String clientIp = socket.getInetAddress().getHostAddress();

                        System.out.println("收到连接请求来自：" + clientIp);

                        new Thread(new Runnable() {
                            @Override
                            public void run() {
                                handleConnection(socket, clientIp);
                            }
                        }).start();
                    }

                } catch (Exception e) {
                    if (isRunning) {
                        e.printStackTrace();
                    }
                }
            }
        }).start();
    }

    public static void stopServer() {
        isRunning = false;

        try {
            if (serverSocket != null && !serverSocket.isClosed()) {
                serverSocket.close();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static void handleConnection(Socket socket, String clientIp) {

        try {

            DataInputStream in = new DataInputStream(socket.getInputStream());

            DataOutputStream out = new DataOutputStream(socket.getOutputStream());

            String command = in.readUTF();

            System.out.println("收到命令：" + command + " 来自：" + clientIp);

            boolean authorized = true;

            if (!authManager.isTrusted(clientIp)) {

                authorized = false;

                if (authCallback != null) {
                    authorized = authCallback.onAuthRequest(clientIp);
                }

                if (authorized) {
                    authManager.addTrustedDevice(clientIp);
                    System.out.println("已授权设备：" + clientIp);
                } else {
                    System.out.println("拒绝设备连接：" + clientIp);
                }
            } else {
                System.out.println("信任设备已连接：" + clientIp);
            }

            if (!authorized) {
                out.writeUTF("DENIED");
                out.flush();
                in.close();
                out.close();
                socket.close();
                return;
            }

            if ("AUTH".equals(command)) {

                out.writeUTF("OK");
                out.writeUTF(UdpListener.getThisDeviceName());
                out.writeUTF(UdpListener.getThisPlatform());
                out.flush();

                System.out.println("设备 " + clientIp + " 连接授权完成");

            } else if ("FILE".equals(command)) {

                out.writeUTF("OK");
                out.flush();

                receiveFiles(in, clientIp);

            } else if ("LIST".equals(command)) {
                out.writeUTF("OK");
                out.flush();
                String path = normalizeRemotePath(in.readUTF());
                String json = listDirectory(path);
                out.writeUTF(json);
                out.flush();

            } else if ("DRIVES".equals(command)) {
                out.writeUTF("OK");
                out.flush();
                String json = listDrives();
                out.writeUTF(json);
                out.flush();

            } else if ("DELETE".equals(command)) {
                out.writeUTF("OK");
                out.flush();
                String targetPath = normalizeRemotePath(in.readUTF());
                boolean success = false;
                try {
                    if (isSafePath(targetPath)) {
                        File f = new File(targetPath);
                        success = f.delete();
                        if (f.isDirectory() && !success) {
                            success = deleteDir(f);
                        }
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
                out.writeUTF(success ? "OK" : "FAILED");
                out.flush();

            } else if ("MKDIR".equals(command)) {
                out.writeUTF("OK");
                out.flush();
                String dirPath = normalizeRemotePath(in.readUTF());
                boolean success = false;
                try {
                    if (isSafePath(dirPath)) {
                        success = new File(dirPath).mkdirs();
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
                out.writeUTF(success ? "OK" : "FAILED");
                out.flush();

            } else if ("RENAME".equals(command)) {
                out.writeUTF("OK");
                out.flush();
                String oldPath = normalizeRemotePath(in.readUTF());
                String newPath = normalizeRemotePath(in.readUTF());
                boolean success = false;
                try {
                    if (isSafePath(oldPath) && isSafePath(newPath)) {
                        success = new File(oldPath).renameTo(new File(newPath));
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
                out.writeUTF(success ? "OK" : "FAILED");
                out.flush();

            } else if ("MOVE".equals(command)) {
                out.writeUTF("OK");
                out.flush();
                String src = normalizeRemotePath(in.readUTF());
                String dest = normalizeRemotePath(in.readUTF());
                boolean success = false;
                try {
                    if (isSafePath(src) && isSafePath(dest)) {
                        Files.move(new File(src).toPath(), new File(dest).toPath(),
                                StandardCopyOption.REPLACE_EXISTING);
                        success = true;
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
                out.writeUTF(success ? "OK" : "FAILED");
                out.flush();

            } else if ("COPY".equals(command)) {
                out.writeUTF("OK");
                out.flush();
                String src = normalizeRemotePath(in.readUTF());
                String dest = normalizeRemotePath(in.readUTF());
                boolean success = false;
                try {
                    if (isSafePath(src) && isSafePath(dest)) {
                        Files.copy(new File(src).toPath(), new File(dest).toPath(),
                                StandardCopyOption.REPLACE_EXISTING);
                        success = true;
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
                out.writeUTF(success ? "OK" : "FAILED");
                out.flush();

            } else if ("DOWNLOAD".equals(command)) {
                String filePath = normalizeRemotePath(in.readUTF());
                File file = new File(filePath);
                if (file.exists() && file.isFile() && isSafePath(filePath)) {
                    out.writeUTF("OK");
                    out.writeLong(file.length());
                    out.flush();
                    FileInputStream fileIn = new FileInputStream(file);
                    byte[] buffer = new byte[8192];
                    int len;
                    while ((len = fileIn.read(buffer)) != -1) {
                        out.write(buffer, 0, len);
                    }
                    fileIn.close();
                } else {
                    out.writeUTF("FAILED");
                }
                out.flush();

            } else {
                out.writeUTF("DENIED");
                out.flush();
                System.out.println("未知请求类型：" + command);
            }

            in.close();
            out.close();
            socket.close();

        } catch (Exception e) {
            System.err.println("处理设备 " + clientIp + " 连接时出错：" + e.getMessage());
            e.printStackTrace();
        }
    }

    private static String normalizeRemotePath(String path) {
        if (path == null || path.length() == 0) {
            return System.getProperty("user.home");
        }

        String normalized = path.replace("\\", "/");

        if ("~".equals(normalized)) {
            return System.getProperty("user.home");
        }

        if (normalized.startsWith("~/")) {
            return System.getProperty("user.home") + normalized.substring(1);
        }

        while (normalized.contains("//")) {
            normalized = normalized.replace("//", "/");
        }

        return normalized;
    }

    private static boolean isSafePath(String path) {
        String lower = normalizeRemotePath(path).toLowerCase();
        String[] forbidden = {
                "c:/windows",
                "c:/program files",
                "c:/program files (x86)",
                "system32"
        };
        for (String f : forbidden) {
            if (lower.contains(f))
                return false;
        }
        return true;
    }

    private static boolean deleteDir(File dir) {
        File[] contents = dir.listFiles();
        if (contents != null) {
            for (File f : contents) {
                if (!deleteDir(f))
                    return false;
            }
        }
        return dir.delete();
    }

    private static String sanitizeRelativePath(String path) {

        if (path == null || path.trim().length() == 0) {
            return "unnamed_file";
        }

        String normalized = path.replace("\\", "/");

        while (normalized.startsWith("/")) {
            normalized = normalized.substring(1);
        }

        String[] parts = normalized.split("/");

        StringBuilder safePath = new StringBuilder();

        for (String part : parts) {

            if (part == null || part.trim().length() == 0) {
                continue;
            }

            if (".".equals(part) || "..".equals(part)) {
                continue;
            }

            part = part.replace(":", "_")
                    .replace("*", "_")
                    .replace("?", "_")
                    .replace("\"", "_")
                    .replace("<", "_")
                    .replace(">", "_")
                    .replace("|", "_");

            if (safePath.length() > 0) {
                safePath.append(File.separator);
            }

            safePath.append(part);
        }

        if (safePath.length() == 0) {
            return "unnamed_file";
        }

        return safePath.toString();
    }

    private static File buildUniqueSaveFile(File originalFile) {

        if (!originalFile.exists()) {
            return originalFile;
        }

        File parent = originalFile.getParentFile();

        String name = originalFile.getName();

        String baseName = name;

        String ext = "";

        int dotIndex = name.lastIndexOf(".");

        if (dotIndex != -1) {
            baseName = name.substring(0, dotIndex);
            ext = name.substring(dotIndex);
        }

        int count = 1;

        File saveFile;

        do {
            saveFile = new File(parent, baseName + "(" + count + ")" + ext);
            count++;
        } while (saveFile.exists());

        return saveFile;
    }

    private static void receiveFiles(DataInputStream in, String clientIp) {

        try {

            int fileCount = in.readInt();

            System.out.println("从 " + clientIp + " 接收文件数量：" + fileCount);

            for (int i = 0; i < fileCount; i++) {

                String relativePath = in.readUTF();

                String safeRelativePath = sanitizeRelativePath(relativePath);

                long fileSize = in.readLong();

                System.out.println(
                        "接收文件：" +
                                safeRelativePath +
                                " 大小：" +
                                formatFileSize(fileSize));

                File receiveDir = new File(AppSettings.getReceiveDir());

                if (!receiveDir.exists()) {
                    receiveDir.mkdirs();
                }

                File saveFile = new File(receiveDir, safeRelativePath);

                File parentFolder = saveFile.getParentFile();

                if (parentFolder != null && !parentFolder.exists()) {
                    parentFolder.mkdirs();
                }

                saveFile = buildUniqueSaveFile(saveFile);

                String savePath = saveFile.getAbsolutePath();

                File tempFile = new File(savePath + ".part");

                if (tempFile.exists()) {
                    tempFile.delete();
                }

                if (receiveCallback != null) {
                    receiveCallback.onFileReceiveStart(
                            clientIp,
                            safeRelativePath,
                            savePath,
                            fileSize);
                }

                FileOutputStream fileOut = new FileOutputStream(tempFile);

                byte[] buffer = new byte[8192];

                long received = 0;

                int len;

                int lastProgress = -1;

                while (received < fileSize) {

                    len = in.read(
                            buffer,
                            0,
                            (int) Math.min(buffer.length, fileSize - received));

                    if (len == -1) {
                        break;
                    }

                    fileOut.write(buffer, 0, len);

                    received += len;

                    int progress;

                    if (fileSize == 0) {
                        progress = 100;
                    } else {
                        progress = (int) ((received * 100) / fileSize);
                    }

                    if (progress != lastProgress) {
                        lastProgress = progress;

                        if (receiveCallback != null) {
                            receiveCallback.onFileReceiveProgress(
                                    clientIp,
                                    safeRelativePath,
                                    savePath,
                                    fileSize,
                                    received,
                                    progress);
                        }
                    }
                }

                fileOut.close();

                if (received == fileSize) {

                    Files.move(
                            tempFile.toPath(),
                            saveFile.toPath(),
                            StandardCopyOption.REPLACE_EXISTING);

                    System.out.println(safeRelativePath + " 接收完成");
                    System.out.println("实际保存路径：" + saveFile.getAbsolutePath());
                    System.out.println("实际接收大小：" + formatFileSize(received));

                    if (receiveCallback != null) {
                        receiveCallback.onFileReceived(
                                clientIp,
                                safeRelativePath,
                                saveFile.getAbsolutePath(),
                                received);
                    }

                } else {

                    if (tempFile.exists()) {
                        tempFile.delete();
                    }

                    System.out.println(
                            safeRelativePath +
                                    " 接收失败：文件未完整接收，已删除临时文件");
                }
            }

            System.out.println("设备 " + clientIp + " 所有文件接收完成\n");

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static String listDirectory(String path) {
        String normalizedPath = normalizeRemotePath(path);
        File dir = new File(normalizedPath);

        System.out.println("LIST 请求路径：" + path);
        System.out.println("LIST 规范路径：" + normalizedPath);
        System.out.println("exists=" + dir.exists() + ", isDirectory=" + dir.isDirectory() + ", canRead=" + dir.canRead());

        StringBuilder json = new StringBuilder();
        json.append("{\"path\":\"").append(escapeJson(normalizedPath)).append("\",\"files\":[");

        boolean first = true;
        int addedCount = 0;
        int maxCount = 300;          // 防止目录内容太多，writeUTF 超过 64KB 后客户端一直转圈
        int maxJsonBytes = 58000;    // DataOutputStream.writeUTF 上限约 65535 bytes，留一点安全余量

        try {
            if (dir.exists() && dir.isDirectory()) {
                File[] files = dir.listFiles();

                if (files != null) {
                    Arrays.sort(files, (a, b) -> {
                        try {
                            if (a.isDirectory() && !b.isDirectory())
                                return -1;
                            if (!a.isDirectory() && b.isDirectory())
                                return 1;
                            return a.getName().compareToIgnoreCase(b.getName());
                        } catch (Exception e) {
                            return 0;
                        }
                    });

                    for (File f : files) {
                        try {
                            if (addedCount >= maxCount) {
                                System.out.println("LIST 提前截断：文件数量超过 " + maxCount + "，路径：" + normalizedPath);
                                break;
                            }

                            String name = f.getName();
                            if (name == null || name.length() == 0) {
                                continue;
                            }

                            if (f.isHidden() || name.startsWith("$")) {
                                continue;
                            }

                            String modified;
                            try {
                                modified = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
                                        .format(new Date(f.lastModified()));
                            } catch (Exception e) {
                                modified = "";
                            }

                            StringBuilder item = new StringBuilder();
                            if (!first) {
                                item.append(",");
                            }
                            item.append("{");
                            item.append("\"name\":\"").append(escapeJson(name)).append("\",");
                            item.append("\"isDir\":").append(f.isDirectory()).append(",");
                            item.append("\"size\":").append(f.isFile() ? f.length() : 0).append(",");
                            item.append("\"modified\":\"").append(modified).append("\"");
                            item.append("}");

                            int nextBytes = (json.length() + item.length() + 2) * 3;
                            if (nextBytes > maxJsonBytes) {
                                System.out.println("LIST 提前截断：JSON 过大，避免 writeUTF 超限。路径：" + normalizedPath);
                                break;
                            }

                            json.append(item);
                            first = false;
                            addedCount++;

                        } catch (Exception fileException) {
                            System.out.println("跳过无法读取的文件项：" + f.getAbsolutePath() + "，原因：" + fileException.getMessage());
                        }
                    }

                    System.out.println("LIST 返回数量：" + addedCount + "，路径：" + normalizedPath);
                } else {
                    System.out.println("LIST 失败：目录无法读取，可能是 macOS 权限未授予 Terminal / Java。路径：" + normalizedPath);
                }
            } else {
                System.out.println("LIST 失败：路径不存在或不是目录：" + normalizedPath);
            }
        } catch (Exception e) {
            System.out.println("LIST 异常，但仍返回空列表。路径：" + normalizedPath + "，原因：" + e.getMessage());
        }

        json.append("]}");

        int bytes = json.toString().getBytes(StandardCharsets.UTF_8).length;
        System.out.println("LIST JSON 大小：" + bytes + " bytes");

        return json.toString();
    }

    private static String listDrives() {
        StringBuilder json = new StringBuilder();
        json.append("{\"drives\":[");

        boolean first = true;
        String os = System.getProperty("os.name").toLowerCase();

        if (os.contains("mac")) {
            String home = System.getProperty("user.home");
            String[] macRoots = {
                    home,
                    home + "/Desktop",
                    home + "/Documents",
                    home + "/Downloads",
                    "/Users",
                    "/Volumes",
                    "/"
            };

            for (String rootPath : macRoots) {
                File root = new File(rootPath);
                if (!root.exists()) {
                    continue;
                }
                if (!first)
                    json.append(",");
                first = false;
                String path = root.getAbsolutePath().replace("\\", "/");
                json.append("\"").append(escapeJson(path)).append("\"");
            }
        } else {
            File[] roots = File.listRoots();
            for (File root : roots) {
                if (!first)
                    json.append(",");
                first = false;
                String path = root.getAbsolutePath().replace("\\", "/");
                json.append("\"").append(escapeJson(path)).append("\"");
            }
        }

        json.append("]}");
        return json.toString();
    }

    private static String escapeJson(String s) {
        if (s == null)
            return "";
        return s.replace("\\", "\\\\").replace("\"", "\\\"");
    }
}

File: linkdisk_flutter/副本lib/network/TcpServer.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

// 依赖你之前已有的类
// import 'AuthManager.dart';
// import 'AppSettings.dart';
// import 'UdpListener.dart';

class TcpServer {
  static AuthManager? _authManager;
  static AuthCallback? _authCallback;
  static ReceiveCallback? _receiveCallback;
  static ServerSocket? _serverSocket;
  static bool _isRunning = false;

  // 启动服务（重载1）
  static void startServer(AuthCallback callback) {
    startServerWithReceive(callback, null);
  }

  // 启动服务（重载2）
  static void startServerWithReceive(
    AuthCallback callback,
    ReceiveCallback? receiveCallback,
  ) {
    if (_isRunning) return;

    _authManager = AuthManager();
    _authCallback = callback;
    _receiveCallback = receiveCallback;
    _isRunning = true;

    // 创建接收目录
    final dir = Directory(AppSettings.getReceiveDir());
    if (!dir.existsSync()) dir.createSync(recursive: true);

    // 启动服务
    _startServerSocket();
  }

  // 停止服务
  static void stopServer() {
    _isRunning = false;
    try {
      _serverSocket?.close();
      _serverSocket = null;
    } catch (e) {
      print(e);
    }
  }

  // 启动 socket 监听
  static Future<void> _startServerSocket() async {
    try {
      _serverSocket = await ServerSocket.bind(InternetAddress.anyIPv4, 6000);
      print("TCP服务器已启动，端口：6000");

      await for (final socket in _serverSocket!) {
        if (!_isRunning) break;

        final clientIp = socket.remoteAddress.address;
        print("收到连接请求来自：$clientIp");

        // 每客户端一个独立隔离处理
        _handleConnection(socket, clientIp);
      }
    } catch (e) {
      if (_isRunning) print(e);
    }
  }

  // 处理单个客户端连接
  static Future<void> _handleConnection(
    Socket socket,
    String clientIp,
  ) async {
    try {
      final inStream = socket;
      final outStream = socket;
      final utf8Decoder = Utf8Decoder();

      // 读取命令 AUTH / FILE
      final command = await _readUtf(inStream);

      // 授权检查
      bool authorized = true;
      if (!_authManager!.isTrusted(clientIp)) {
        authorized = false;
        if (_authCallback != null) {
          authorized = _authCallback!.onAuthRequest(clientIp);
        }
        if (authorized) {
          _authManager!.addTrustedDevice(clientIp);
          print("已授权设备：$clientIp");
        } else {
          print("拒绝设备连接：$clientIp");
        }
      } else {
        print("信任设备已连接：$clientIp");
      }

      // 未授权直接拒绝
      if (!authorized) {
        _writeUtf(outStream, "DENIED");
        await socket.flush();
        await socket.close();
        return;
      }

      // 处理 AUTH
      if (command == "AUTH") {
        _writeUtf(outStream, "OK");
        _writeUtf(outStream, UdpListener.getThisDeviceName());
        _writeUtf(outStream, UdpListener.getThisPlatform());
        await outStream.flush();
        print("设备 $clientIp 连接授权完成");
      }

      // 处理 FILE
      else if (command == "FILE") {
        _writeUtf(outStream, "OK");
        await outStream.flush();
        await _receiveFiles(inStream, clientIp);
      }

      // 未知命令
      else {
        _writeUtf(outStream, "DENIED");
        await outStream.flush();
        print("未知请求类型：$command");
      }

      await socket.close();
    } catch (e) {
      print("处理设备 $clientIp 连接时出错：${e.toString()}");
      await socket.close();
    }
  }

  // ------------------------------
  // 接收文件核心逻辑
  // ------------------------------
  static Future<void> _receiveFiles(Socket socket, String clientIp) async {
    try {
      // 读取文件数量
      final fileCount = await _readInt(socket);
      print("从 $clientIp 接收文件数量：$fileCount");

      final receiveDir = Directory(AppSettings.getReceiveDir());
      if (!receiveDir.existsSync()) {
        receiveDir.createSync(recursive: true);
      }

      for (int i = 0; i < fileCount; i++) {
        // 读取相对路径 & 文件大小
        final relativePath = await _readUtf(socket);
        final fileSize = await _readLong(socket);

        // 安全路径处理
        final safePath = _sanitizeRelativePath(relativePath);
        print("接收文件：$safePath 大小：${_formatFileSize(fileSize)}");

        // 构建保存路径
        final saveFile = File('${receiveDir.path}${Platform.pathSeparator}$safePath');
        final parent = saveFile.parent;
        if (!parent.existsSync()) parent.createSync(recursive: true);

        // 重命名避免覆盖
        final uniqueFile = _buildUniqueSaveFile(saveFile);
        final savePath = uniqueFile.path;

        // 临时文件
        final tempFile = File('$savePath.part');
        if (tempFile.existsSync()) await tempFile.delete();

        // 开始回调
        _receiveCallback?.onFileReceiveStart(
          clientIp,
          safePath,
          savePath,
          fileSize,
        );

        // 写入文件
        final fileOut = tempFile.openWrite();
        final buffer = Uint8List(8192);
        int received = 0;
        int lastProgress = -1;

        while (received < fileSize) {
          final need = (buffer.length < (fileSize - received))
              ? buffer.length
              : (fileSize - received);

          final bytes = await socket.read(need);
          if (bytes == null || bytes.isEmpty) break;

          fileOut.add(bytes);
          await fileOut.flush();

          received += bytes.length;

          // 进度
          final progress = (fileSize == 0)
              ? 100
              : (received * 100 ~/ fileSize);

          if (progress != lastProgress) {
            lastProgress = progress;
            _receiveCallback?.onFileReceiveProgress(
              clientIp,
              safePath,
              savePath,
              fileSize,
              received,
              progress,
            );
          }
        }

        await fileOut.close();

        // 接收完成
        if (received == fileSize) {
          await tempFile.rename(savePath);
          print("$safePath 接收完成");
          print("实际保存路径：$savePath");
          _receiveCallback?.onFileReceived(
            clientIp,
            safePath,
            savePath,
            received,
          );
        } else {
          if (await tempFile.exists()) await tempFile.delete();
          print("$safePath 接收失败：文件不完整");
        }
      }

      print("设备 $clientIp 所有文件接收完成\n");
    } catch (e) {
      print(e);
    }
  }

  // ------------------------------
  // 工具：路径安全处理（防路径穿越）
  // ------------------------------
  static String _sanitizeRelativePath(String path) {
    if (path.trim().isEmpty) return "unnamed_file";

    var normalized = path.replaceAll('\\', '/');
    while (normalized.startsWith('/')) {
      normalized = normalized.substring(1);
    }

    final parts = normalized.split('/');
    final safe = StringBuffer();

    for (final part in parts) {
      if (part.isEmpty || part == '.' || part == '..') continue;

      var clean = part
          .replaceAll(':', '_')
          .replaceAll('*', '_')
          .replaceAll('?', '_')
          .replaceAll('"', '_')
          .replaceAll('<', '_')
          .replaceAll('>', '_')
          .replaceAll('|', '_');

      if (safe.isNotEmpty) safe.write(Platform.pathSeparator);
      safe.write(clean);
    }

    return safe.isEmpty ? "unnamed_file" : safe.toString();
  }

  // ------------------------------
  // 工具：自动重命名文件（避免覆盖）
  // ------------------------------
  static File _buildUniqueSaveFile(File file) {
    if (!file.existsSync()) return file;

    final name = file.path.split(Platform.pathSeparator).last;
    final dir = file.parent.path;

    final dot = name.lastIndexOf('.');
    var base = name;
    var ext = '';

    if (dot != -1) {
      base = name.substring(0, dot);
      ext = name.substring(dot);
    }

    int count = 1;
    while (true) {
      final newPath = '$dir${Platform.pathSeparator}$base($count)$ext';
      final f = File(newPath);
      if (!f.existsSync()) return f;
      count++;
    }
  }

  // ------------------------------
  // 工具：文件大小格式化
  // ------------------------------
  static String _formatFileSize(int size) {
    if (size < 1000) return '$size B';
    double v = size / 1000;
    if (v < 1000) return '${v.toStringAsFixed(2)} KB';
    v /= 1000;
    if (v < 1000) return '${v.toStringAsFixed(2)} MB';
    v /= 1000;
    return '${v.toStringAsFixed(2)} GB';
  }

  // ------------------------------
  // 底层读写工具（对应 Java DataInputStream）
  // ------------------------------
  static Future<String> _readUtf(Socket socket) async {
    final line = await utf8.decodeStream(socket);
    return line.trim();
  }

  static void _writeUtf(Socket socket, String text) {
    socket.write('$text\n');
  }

  static Future<int> _readInt(Socket socket) async {
    final b = await _readBytes(socket, 4);
    return ByteData.sublistView(b).getInt32(0, Endian.big);
  }

  static Future<int> _readLong(Socket socket) async {
    final b = await _readBytes(socket, 8);
    return ByteData.sublistView(b).getInt64(0, Endian.big);
  }

  static Future<Uint8List> _readBytes(Socket socket, int len) async {
    final b = BytesBuilder();
    while (b.length < len) {
      final chunk = await socket.first;
      b.add(chunk);
    }
    return b.toBytes().sublist(0, len);
  }
}

// ------------------------------
// 回调接口
// ------------------------------
abstract class AuthCallback {
  bool onAuthRequest(String ip);
}

abstract class ReceiveCallback {
  void onFileReceiveStart(
    String clientIp,
    String fileName,
    String savePath,
    int fileSize,
  );

  void onFileReceiveProgress(
    String clientIp,
    String fileName,
    String savePath,
    int fileSize,
    int receivedBytes,
    int progress,
  );

  void onFileReceived(
    String clientIp,
    String fileName,
    String savePath,
    int fileSize,
  );
}